<a href="https://colab.research.google.com/github/Selvarani93/Assignment_Bluejay/blob/main/Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd

In [ ]:
df = pd.read_csv('/content/Assignment_Timecard.xlsx - Sheet1.csv')

In [ ]:
df

Position ID Position Status                 Time             Time Out  \
0      WFS000054          Active                  NaN                  NaN   
1      WFS000065          Active  09/12/2023 10:08 AM  09/12/2023 01:53 PM   
2      WFS000065          Active  09/12/2023 02:23 PM  09/12/2023 07:02 PM   
3      WFS000065          Active  09/13/2023 10:08 AM  09/13/2023 02:20 PM   
4      WFS000065          Active  09/13/2023 02:50 PM  09/13/2023 08:44 PM   
...          ...             ...                  ...                  ...   
1479   WFS000589          Active  09/20/2023 09:55 AM  09/20/2023 02:30 PM   
1480   WFS000589          Active  09/20/2023 03:00 PM  09/20/2023 07:29 PM   
1481   WFS000589          Active  09/21/2023 09:56 AM  09/21/2023 02:30 PM   
1482   WFS000589          Active  09/21/2023 03:00 PM  09/21/2023 07:16 PM   
1483   WFS000591          Active                  NaN                  NaN   

     Timecard Hours (as Time) Pay Cycle Start Date Pay Cycle End Date  \
0                         NaN                  NaN                NaN   
1                        3:45           09/10/2023         09/23/2023   
2                        4:39           09/10/2023         09/23/2023   
3                        4:12           09/10/2023         09/23/2023   
4                        5:54           09/10/2023         09/23/2023   
...                       ...                  ...                ...   
1479                     4:35           09/10/2023         09/23/2023   
1480                     4:29           09/10/2023         09/23/2023   
1481                     4:34           09/10/2023         09/23/2023   
1482                     4:16           09/10/2023         09/23/2023   
1483                      NaN                  NaN                NaN   

                    Employee Name  File Number  
0                 SiWgh, PraGhjEM           54  
1                 REsaXiaWE, XAis           65  
2                 REsaXiaWE, XAis           65  
3                 REsaXiaWE, XAis           65  
4                 REsaXiaWE, XAis           65  
...                           ...          ...  
1479              WgAyeW, RayCEWd          589  
1480              WgAyeW, RayCEWd          589  
1481              WgAyeW, RayCEWd          589  
1482              WgAyeW, RayCEWd          589  
1483  ArveXE RECerE, AWdres JesAs          591  

[1484 rows x 9 columns]

Data has 2 unnamed columns, we'll remove them.
Column 10 isn't empty but has unwanted information. So, removing that column using drop.

In [ ]:
df.dropna(axis=1, how='all', inplace=True)
print(df)

     Position ID Position Status                 Time             Time Out  \
0      WFS000054          Active                  NaN                  NaN   
1      WFS000065          Active  09/12/2023 10:08 AM  09/12/2023 01:53 PM   
2      WFS000065          Active  09/12/2023 02:23 PM  09/12/2023 07:02 PM   
3      WFS000065          Active  09/13/2023 10:08 AM  09/13/2023 02:20 PM   
4      WFS000065          Active  09/13/2023 02:50 PM  09/13/2023 08:44 PM   
...          ...             ...                  ...                  ...   
1479   WFS000589          Active  09/20/2023 09:55 AM  09/20/2023 02:30 PM   
1480   WFS000589          Active  09/20/2023 03:00 PM  09/20/2023 07:29 PM   
1481   WFS000589          Active  09/21/2023 09:56 AM  09/21/2023 02:30 PM   
1482   WFS000589          Active  09/21/2023 03:00 PM  09/21/2023 07:16 PM   
1483   WFS000591          Active                  NaN                  NaN   

     Timecard Hours (as Time) Pay Cycle Start Date Pay Cycle En

In [ ]:
df.isnull().any()

Position ID                 False
Position Status             False
Time                         True
Time Out                     True
Timecard Hours (as Time)     True
Pay Cycle Start Date         True
Pay Cycle End Date           True
Employee Name               False
File Number                 False
dtype: bool

In [ ]:
df[df['Time'].isna()==1]

Position ID Position Status Time Time Out Timecard Hours (as Time)  \
0      WFS000054          Active  NaN      NaN                      NaN   
89     WFS000174          Active  NaN      NaN                      NaN   
158    WFS000196          Active  NaN      NaN                      NaN   
231    WFS000256          Active  NaN      NaN                      NaN   
535    WFS000432          Active  NaN      NaN                      NaN   
552    WFS000436          Active  NaN      NaN                      NaN   
1056   WFS000529          Active  NaN      NaN                      NaN   
1066   WFS000532          Active  NaN      NaN                      NaN   
1199   WFS000553          Active  NaN      NaN                      NaN   
1483   WFS000591          Active  NaN      NaN                      NaN   

     Pay Cycle Start Date Pay Cycle End Date                Employee Name  \
0                     NaN                NaN              SiWgh, PraGhjEM   
89                    NaN                NaN              MErres, EdAardE   
158                   NaN                NaN                XEpes, JerrEd   
231                   NaN                NaN                 Ware, JEshAa   
535                   NaN                NaN               KaAr, ArviWder   
552                   NaN                NaN            WiXGard, Shadrack   
1056                  NaN                NaN        PearsEW, DeXviW DevEW   
1066                  NaN                NaN               REiz, CarviW E   
1199                  NaN                NaN     SEMEsEXis, Richar SMeveW   
1483                  NaN                NaN  ArveXE RECerE, AWdres JesAs   

      File Number  
0              54  
89            174  
158           196  
231           256  
535           432  
552           436  
1056          529  
1066          532  
1199          553  
1483          591

In [ ]:
df['Time'].fillna(method='ffill', inplace=True)

In [ ]:
df['Time Out'].fillna(method='ffill', inplace=True)

In [ ]:
df['Timecard Hours (as Time)'].fillna(method='ffill', inplace=True)

In [ ]:
df['Pay Cycle Start Date'].fillna(method='ffill', inplace=True)

In [ ]:
df['Pay Cycle End Date'].fillna(method='ffill', inplace=True)

In [ ]:
df.isnull().any()

Position ID                 False
Position Status             False
Time                        False
Time Out                    False
Timecard Hours (as Time)    False
Pay Cycle Start Date        False
Pay Cycle End Date          False
Employee Name               False
File Number                 False
dtype: bool

In [ ]:
df[df['Timecard Hours (as Time)'].isna()==1]

Position ID Position Status Time Time Out Timecard Hours (as Time)  \
0   WFS000054          Active    0        0                      NaN   

  Pay Cycle Start Date Pay Cycle End Date    Employee Name  File Number  
0                  NaN                NaN  SiWgh, PraGhjEM           54

In [ ]:
df[df['Pay Cycle Start Date'].isna()==1]

Position ID Position Status Time Time Out Timecard Hours (as Time)  \
0   WFS000054          Active    0        0                      NaN   

  Pay Cycle Start Date Pay Cycle End Date    Employee Name  File Number  
0                  NaN                NaN  SiWgh, PraGhjEM           54

In [ ]:
#df.drop(index=0, inplace=True) //executed this to get rid of the 1st record as it is only one record. so it is fine to drop it.

In [ ]:
df.isnull().any()

Position ID                 False
Position Status             False
Time                        False
Time Out                    False
Timecard Hours (as Time)    False
Pay Cycle Start Date        False
Pay Cycle End Date          False
Employee Name               False
File Number                 False
dtype: bool

Now, we do not have any NaN values, and everything has been filled.

In [ ]:
df.duplicated()

1       False
2       False
3       False
4       False
5       False
        ...  
1479    False
1480    False
1481    False
1482    False
1483    False
Length: 1483, dtype: bool

There are no duplicate records. So, no need to remove the duplicates.

Just to be sure, let's check if the formatting of time is correct or not.

In [ ]:
try:
    df['Time'] = pd.to_datetime(df['Time'])
    print("Date format is correct.")
except ValueError as e:
    print(f"Error: {e}. Date format might be incorrect.")

Date format is correct.


In [ ]:
try:
    df['Time Out'] = pd.to_datetime(df['Time Out'])
    print("Time format is correct.")
except ValueError as e:
    print(f"Error: {e}. Time format might be incorrect.")

Time format is correct.


In [ ]:
df['Employee Name'] = df['Employee Name'].str.title()

here, the first letter of each word in the employee name column has been capitalized and formatted in a neat way.

In [ ]:
df.head()

Position ID Position Status                Time            Time Out  \
1   WFS000065          Active 2023-09-12 10:08:00 2023-09-12 13:53:00   
2   WFS000065          Active 2023-09-12 14:23:00 2023-09-12 19:02:00   
3   WFS000065          Active 2023-09-13 10:08:00 2023-09-13 14:20:00   
4   WFS000065          Active 2023-09-13 14:50:00 2023-09-13 20:44:00   
5   WFS000065          Active 2023-09-14 10:09:00 2023-09-14 14:30:00   

  Timecard Hours (as Time) Pay Cycle Start Date Pay Cycle End Date  \
1                     3:45           09/10/2023         09/23/2023   
2                     4:39           09/10/2023         09/23/2023   
3                     4:12           09/10/2023         09/23/2023   
4                     5:54           09/10/2023         09/23/2023   
5                     4:21           09/10/2023         09/23/2023   

     Employee Name  File Number  
1  Resaxiawe, Xais           65  
2  Resaxiawe, Xais           65  
3  Resaxiawe, Xais           65  
4  Resaxiawe, Xais           65  
5  Resaxiawe, Xais           65

In [ ]:
df.sort_values(by=['Employee Name', 'Time'], inplace=True)

In [ ]:
df['Day Diff'] = df.groupby('Employee Name')['Time'].diff().dt.days

In [ ]:
consecutive_days_df = df[df['Day Diff'] == 1].groupby('Employee Name').filter(lambda x: len(x) >= 7)

In [ ]:
df['Employee Name'] = df['Employee Name'].str.strip().str.title()

In [ ]:
result = consecutive_days_df[['Employee Name', 'Position ID']].drop_duplicates()
print(result)

Empty DataFrame
Columns: [Employee Name, Position ID]
Index: []


**THERE ARE NO EMPLOYEE WHO WORKED FOR 7 CONSECUTIVE DAYS.**

In [ ]:
df['Employee Name'].isnull().sum()

0

In [ ]:
grouped_df = df.groupby('Employee Name').mean()
print(grouped_df)

                          File Number  Day Diff
Employee Name                                  
Agex, Wamhaw Regerm Pewa        588.0  0.444444
Arcewma Fxeres, Jewamhaw        584.0  0.176471
Aresmigai, Axexawder            306.0  0.333333
Arevaxe, Jese                   491.0  0.294118
Arias, Fexipe                   170.0  0.176471
...                               ...       ...
Xepes, Jerred                   196.0       NaN
Xepez Wawez, Xais               522.0  0.235294
Xiww, Jasmiw                    462.0  0.428571
Zacera, Cigaex                  499.0  0.111111
Zacera, Edward                  477.0  0.500000

[105 rows x 2 columns]


<ipython-input-71-295a7f73d212>:1: FutureWarning: The default value of numeric_only in DataFrameGroupBy.mean is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  grouped_df = df.groupby('Employee Name').mean()


In [ ]:
df['Time'] = pd.to_datetime(df['Time'])
df['Date Diff'] = df.groupby('Employee Name')['Time'].diff().dt.days
print(df[['Employee Name', 'Time', 'Date Diff']])

                 Employee Name                Time  Date Diff
1463  Agex, Wamhaw Regerm Pewa 2023-09-13 10:40:00        NaN
1464  Agex, Wamhaw Regerm Pewa 2023-09-13 15:30:00        0.0
1465  Agex, Wamhaw Regerm Pewa 2023-09-18 09:55:00        4.0
1466  Agex, Wamhaw Regerm Pewa 2023-09-18 14:29:00        0.0
1467  Agex, Wamhaw Regerm Pewa 2023-09-19 10:25:00        0.0
...                        ...                 ...        ...
722             Zacera, Edward 2023-09-15 14:30:00        0.0
723             Zacera, Edward 2023-09-17 10:41:00        1.0
724             Zacera, Edward 2023-09-17 15:30:00        0.0
725             Zacera, Edward 2023-09-20 10:40:00        2.0
726             Zacera, Edward 2023-09-20 16:00:00        0.0

[1483 rows x 3 columns]


In [ ]:
consecutive_days_df = df[df['Date Diff'] == 1].groupby('Employee Name').filter(lambda x: len(x) >= 5)
#checking if any employee worked for 5 consecutive days


In [ ]:
result = consecutive_days_df[['Employee Name', 'Position ID']].drop_duplicates()
print(result)

    Employee Name Position ID
638  Xiww, Jasmiw   WFS000462


index 638 has a record of working for 5 consecutive days.
and nobody for 7 consecutive days.
THIS ANSWERS THE FIRST QUESION, THAT NOBODY HAS WORKED FOR 7 CONSECUTIVE DAYS.

In [ ]:
df['Time Diff'] = df.groupby('Employee Name')['Time'].diff()
#calculating difference between consecutive shifts

In [ ]:
filtered_df = df[(df['Time Diff'] > '1:00:00') & (df['Time Diff'] < '10:00:00')]
#filtering the df

In [ ]:
filtered_df[['Employee Name', 'Position ID', 'Position Status']].drop_duplicates()
#Displaying

Employee Name Position ID Position Status
1464  Agex, Wamhaw Regerm Pewa   WFS000588          Active
1409  Arcewma Fxeres, Jewamhaw   WFS000584          Active
250       Aresmigai, Axexawder   WFS000306          Active
728              Arevaxe, Jese   WFS000491          Active
72               Arias, Fexipe   WFS000170          Active
...                        ...         ...             ...
160                 Xee, Xacar   WFS000200          Active
1009         Xepez Wawez, Xais   WFS000522          Active
637               Xiww, Jasmiw   WFS000462          Active
785             Zacera, Cigaex   WFS000499          Active
717             Zacera, Edward   WFS000477          Active

[96 rows x 3 columns]

**THIS ANSWERS THE SECOND QUESTION, who have less than 10 hours of time between shifts but greater than 1 hour**

let's see the first occurences of each employee having this criteria.

In [ ]:
unique_names_first_occurrence_df = filtered_df.groupby('Employee Name').first().reset_index()[['Employee Name', 'Time', 'Time Diff']]
unique_names_first_occurrence_df


Employee Name                Time       Time Diff
0   Agex, Wamhaw Regerm Pewa 2023-09-13 15:30:00 0 days 04:50:00
1   Arcewma Fxeres, Jewamhaw 2023-09-10 15:00:00 0 days 04:50:00
2       Aresmigai, Axexawder 2023-09-11 15:00:00 0 days 04:41:00
3              Arevaxe, Jese 2023-09-10 05:29:00 0 days 03:37:00
4              Arias, Fexipe 2023-09-10 05:57:00 0 days 03:57:00
..                       ...                 ...             ...
91                Xee, Xacar 2023-09-10 10:00:00 0 days 04:05:00
92         Xepez Wawez, Xais 2023-09-10 07:30:00 0 days 04:25:00
93              Xiww, Jasmiw 2023-09-10 15:00:00 0 days 04:43:00
94            Zacera, Cigaex 2023-09-12 15:00:00 0 days 04:48:00
95            Zacera, Edward 2023-09-10 15:00:00 0 days 04:08:00

[96 rows x 3 columns]

In [ ]:
df['Shift Duration'] = df['Time Out'] - df['Time']
#calculating the duration of each shifts

In [ ]:
long_shifts_df = df[df['Shift Duration'] > '14:00:00']
#filtering the df which are greater than 14 hours

In [ ]:
print(long_shifts_df[['Employee Name', 'Time', 'Position ID']])
#displaying

                        Employee Name                Time Position ID
1036  Dexgadixxe Redarme, Chrismiaw S 2023-09-17 04:00:00   WFS000523


**THIS ANSWERS THE THIRD QUESTION,
"Dexgadixxe Redarme" with index = 1036 has worked for more than 14 hours in a single shift.**